projects/kit/components/rating/rating.component.ts

```diff
@@ -0,0 +1,100 @@
+import {
+    ChangeDetectionStrategy,
+    ChangeDetectorRef,
+    Component,
+    ElementRef,
+    forwardRef,
+    HostBinding,
+    Inject,
+    Input,
+    Optional,
+    Self,
+    ViewChild,
+} from '@angular/core';
+import {NgControl} from '@angular/forms';
+import {
+    AbstractTuiControl,
+    isNativeFocused,
+    TUI_FOCUSABLE_ITEM_ACCESSOR,
+} from '@taiga-ui/cdk';
+
+import {RatingOptions, TUI_RATING_OPTIONS} from './rating.options';
+import {roundRatingBy} from './utils/round-rating-by';
+
+@Component({
+    selector: 'tui-rating',
+    templateUrl: './rating.template.html',
+    styleUrls: ['./rating.style.less'],
+    changeDetection: ChangeDetectionStrategy.OnPush,
+    providers: [
+        {
+            provide: TUI_FOCUSABLE_ITEM_ACCESSOR,
+            useExisting: forwardRef(() => TuiRatingComponent),
+        },
+    ],
+})
+export class TuiRatingComponent extends AbstractTuiControl<number> {
+    @ViewChild('focusableElement')
+    private readonly focusableElement?: ElementRef<HTMLInputElement>;
+
+    @Input() max: number = this.options.max;
+    @Input() iconNormal: string = this.options.iconNormal;
+    @Input() iconFilled: string = this.options.iconFilled;
+
+    constructor(
+        @Optional()
+        @Self()
+        @Inject(NgControl)
+        ngControl: NgControl | null,
+        @Inject(ChangeDetectorRef)
+        changeDetectorRef: ChangeDetectorRef,
+        @Inject(TUI_RATING_OPTIONS)
+        private readonly options: RatingOptions,
+    ) {
+        super(ngControl, changeDetectorRef);
+    }
+
+    @HostBinding('class._disabled')
+    get computedDisabled(): boolean {
+        return this.disabled;
+    }
+
+    get nativeFocusableElement(): HTMLInputElement | null {
+        return this.computedDisabled || !this.focusableElement
+            ? null
+            : this.focusableElement.nativeElement;
+    }
+
+    get focused(): boolean {
+        return isNativeFocused(this.nativeFocusableElement);
+    }
+
+    get isFocusable(): boolean {
+        return !(this.readOnly || this.disabled);
+    }
+
+    get percent(): number {
+        return (100 * this.value) / this.max;
+    }
+
+    setRateByIndex(index: number): void {
+        const computed = this.getComputedRatingValue(this.max - index);
+        const toggled = this.value === computed ? 0 : computed;
+
+        this.updateValue(toggled);
+    }
+
+    setRate(rate: number): void {
+        const computed = this.getComputedRatingValue(rate);
+
+        this.updateValue(computed);
+    }
+
+    protected getFallbackValue(): number {
+        return 0;
+    }
+
+    private getComputedRatingValue(value: number): number {
+        return roundRatingBy({value, max: this.max});
```

Comment:
Why do we need this? What is the case when we get invalid value?

Reply:
Because you can set invalid value by `setRate` or `setRateByIndex`

Reply:
fixed